memory.py

```diff
@@ -15,46 +14,71 @@ def __init__(self, size):
     self.index = 0
     self.size = size
     self.full = False  # Used to track actual capacity
-    self.sum_tree = np.zeros((2 * size - 1, ), dtype=np.float32)  # Initialise fixed size tree with all (priority) zeros
-    self.data = np.array([None] * size)  # Wrap-around cyclic buffer
+    self.tree_start = 1
+    while self.tree_start < self.size: # Put all used node leaves on last tree level
+      self.tree_start *= 2
+    self.tree_start -= 1
+    self.sum_tree = np.zeros((self.tree_start + self.size,), dtype=np.float32)
+    self.data = np.array([blank_trans] * size, dtype=Transition_dtype)  # Build structured array
     self.max = 1  # Initial max value to return (1 = 1^ω)
 
-  # Propagates value up tree given a tree index
-  def _propagate(self, index, value):
+  # Updates nodes values from current tree
+  def _update_nodes(self, indices):
+    children_indices = indices * 2 + np.expand_dims([1, 2], axis=1)
+    self.sum_tree[indices] = np.sum(self.sum_tree[children_indices], axis=0)
+
+  # Propagates changes up tree given tree indices
+  def _propagate(self, indices):
+    parents = (indices - 1) // 2
+    unique_parents = np.unique(parents)
+    self._update_nodes(unique_parents)
+    if parents[0] != 0:
+      self._propagate(parents)
+
+  # Propagates single value up tree given a tree index for efficiency
+  def _propagate_index(self, index, value):
```

Comment:
Just wondering the reasoning why you renamed `_propagate` to `_propagate_index` and introduced a different `_propagate`?

Reply:
The main functions to update/propagate values work with multiple indices in a numpy array, that is why I introduced a new `_propagate`. However, when we have to update a single value after appending a new transition of experience, I wanted to avoid the overhead of having to create and process a numpy array with a single index. Therefore I decided to keep also the old implementation under `_propagate_index` (I have now removed the value argument which I noticed was unused).